app/models/storage_location.rb

```diff
@@ -78,6 +81,18 @@ def size
     inventory_items.sum(:quantity)
   end
 
+  def discard
+    return false if size > 0
+
+    super
+  end
+
+  def discard!
+    raise Errors::StorageLocationNotEmpty if size > 0
+
+    super
+  end
```

Comment:
I noticed that this was added inside the model which does offer the benefit of not being able to run `discard` anywhere (in the console) and receive this warning. However, I think there are some benefits to adopting a service object-like approach to encapsulate the logic behind what is or what isn't allowed when it comes to discarding. Something like a `StorageLocations::Deactivate` class which is an isolated unit of code that handles this business logic. 

My suggestion for going in that direction is that I've historically seen in this project that it isn't clear why code is added if it veers too far from where it's being used (in this case its the discarding and it is a controller endpoint)

Let me know what you think :)!

Reply:
Hey, @jacobherrington appreciate your openness! 

Yeh I think something like `RequestDestroyService` sounds god to me. 

> It seems like discard's [callbacks](https://github.com/jhawthorn/discard#callbacks) feature might also be a better approach than what I did here! In that way we are not behaving too differently than the library expects.

This is a slight school of thinking and also against basecamp's style but I generally avoid callbacks. In my experience, it leads to unintended side effects. Not saying it isn't possible to build clean and clear code with callbacks, but in my experience using a service object helps encapsulate the domain and more resilient to tangled code.

Let me know what ya think.